tests/tests_eve/unit_tests/test_datamodels.py

```diff
@@ -0,0 +1,892 @@
+            inherited_model_with_validators_factory(extra_value=True)
+
+        # Regular behavior for new field
+        inherited_model_with_validators_factory(int_value=1, new_int_value=1)
+        with pytest.raises(TypeError, match="new_int_value"):
+            inherited_model_with_validators_factory(int_value=1, new_int_value=1.0)
+        with pytest.raises(ValueError, match="new_int_value"):
+            inherited_model_with_validators_factory(int_value=1, new_int_value=2)
+
+
+class TestRootValidators:
+    def test_root_validators(self, model_with_root_validators_factory):
+        model_with_root_validators_factory()
+
+        model_with_root_validators_factory(int_value=1, str_value="1")
+        with pytest.raises(ValueError, match="float_value"):
+            model_with_root_validators_factory(int_value=1, float_value=1.0)
+
+    def test_inherited_root_validators(self, inherited_model_with_root_validators_factory):
+        inherited_model_with_root_validators_factory()
+
+        with pytest.raises(ValueError, match="str_value"):
+            inherited_model_with_root_validators_factory(int_value=1, str_value="1")
+        with pytest.raises(ValueError, match="float_value"):
+            inherited_model_with_root_validators_factory(int_value=1, float_value=1.0)
+            inherited_model_with_root_validators_factory(
+                int_value=1, float_value=1.0, str_value="1"
+            )
+
+
+class TestGenericModels:
+    @pytest.mark.parametrize("concrete_type", [int, float, str])
+    def test_generic_model_instantiation_name(self, concrete_type):
+        Model = datamodels.concretize(GenericModel, concrete_type)
+        assert concrete_type.__name__ in Model.__name__
+
+    @pytest.mark.parametrize("concrete_type", [int, float, str])
+    def test_generic_model_instantiation_cache(self, concrete_type):
+        Model1 = datamodels.concretize(GenericModel, concrete_type)
+        Model2 = datamodels.concretize(GenericModel, concrete_type)
+        Model3 = datamodels.concretize(GenericModel, concrete_type)
+
+        assert (
+            Model1 is Model2
+            and Model2 is Model3
+            and Model3 is datamodels.concretize(GenericModel, concrete_type)
+        )
+
+    @pytest.mark.parametrize("concrete_type", [int, float, str])
+    def test_generic_model_alias(self, concrete_type):
+        Model = datamodels.concretize(GenericModel, concrete_type)
+
+        assert GenericModel[concrete_type].__class__ is Model
+        assert typing.get_origin(GenericModel[concrete_type]) is Model
+
+        class SubModel(GenericModel[concrete_type]):
+            ...
+
+        assert SubModel.__base__ is Model
+
+    @pytest.mark.parametrize(
+        "value", [False, 1, 1.1, "string", [1], ("a", "b"), {1, 2, 3}, {"a": 1}]
+    )
+    def test_generic_field_type_validation(self, generic_model_factory, value):
+        generic_model_factory(generic_value="")
```

Comment:
Don't understand what's happening here? where is `value` used?

Reply:
oops, it should be used/renamed as `generic_value`

Reply:
Done